### ISSUE TITLE
'scoreDiff is not finite' error in ELBO optimization

### ISSUE THREAD

**Issue by JsonFreeman** [stuhlmueller/webppl-intercache#1]
To reproduce:
1. Check out this repo and switch to branch `bypass`
2. Use the latest webppl from the daipp branch (I am on commit ba04723e - "Reintroduce mapData")
3. From the root of the intercache repo, run

> webppl tests/test3.wppl --require webppl-timeit --require webppl-json --require .

After at most 20 iterations iterations of the cache, you will see the error:

```
ELBO: scoreDiff is not finite.
    at /Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/src/inference/elbo.js:48

47|                 assert.ok(ad.scalar.peq(typeof scoreDiff, 'number'));
48|                 assert.ok(_.isFinite(scoreDiff), 'ELBO: scoreDiff is not finite.');
    -----------------------^
49|                 var useLR = sameAdNode(this.logq, this.logr);

Stack trace:
    at ELBO.<anonymous> (/Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/src/inference/elbo.js:48:24)
    at eval (eval at <anonymous> (/Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/src/main.js:222:10), <anonymous>:497:80)
    at /Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/src/util.js:51:13
    at /Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/src/util.js:14:14
    at eval (eval at <anonymous> (/Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/src/main.js:222:10), <anonymous>:5:13)
    at /Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/src/main.js:222:48
    at Object.timeif (/Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/src/util.js:283:37)
    at Object.run (/Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/src/main.js:221:8)
    at run (/Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/webppl:36:12)
    at main (/Users/jfreeman/Documents/StanfordCourses/GoodmanResearch/webppl/webppl:129:3)
```

Each iteration of the cache involves at most one adagrad ELBO optimization call in the updateParams function in library.wppl. It seems that when `returnValVariance` in `fastFunc` is low, the crash is likely to happen, though it does not make sense to me that these would be connected.

Things I have tried
1. Reducing the adagrad step size from 0.01 to 0.001
2. Using 10 samples instead of 1 for gradient estimation
3. Switch adagrad to gd and adam
4. Clipping gradient
5. All distributions in the optimized function `distOnFuncs` are Gaussians (including the guide distributions).

@null-a do you know what could be the issue here?

**Comment by null-a** [comment]
Oops! I think I might have broken this when I renamed one of the params that `Optimize` takes. We now have `Optimize(model, {optMethod: {...}})` rather than `Optimize(model, {method: {...}})`, which means that the changes you tried making to the optimization method/learning rate were having no effect.

After changing `method -> optMethod` [here](https://github.com/stuhlmueller/webppl-intercache/blob/88f7d0d826b2168e25544b8de54e6da5ecd81e0c/src/library.wppl#L178) `test3.wppl` ran without error[1]. If this has previously worked and then stopped working, this is probably why. Sorry about that!

[1] Actually, it died at the end because `mapReduce1` has been moved out of core webppl and into [`webppl-array-utils`](https://github.com/probmods/webppl-array-utils).

**Comment by JsonFreeman** [comment]
Thank you Paul! `optMethod` seems to have done the trick.

**Comment by JsonFreeman** [comment]
@null-a, this issue seems to come back when I increase the step size from 0.001 (even just to 0.002). Essentially, it seems that no matter what step size I select, once the optimization is close enough to converged, the crash happens if we try to keep learning. Do you know what I could do to keep learning even once the learning has converged?

I'm going to try the gradient clipping and increasing the gradient samples again.

**Comment by null-a** [comment]
Ah, I think I see the problem -- one of the variance params of the guide eventually becomes negative. If you put the [variance params](https://github.com/stuhlmueller/webppl-intercache/blob/ec2730eb8c651422ccc8d0f245b8f6720f010dfb/src/library.wppl#L79) through some squishing function or other to ensure they stay positive I think it works:

``` js
var guideVariance = Math.exp(param(call(options.guide.initialVarianceFunc, i)));
```

If you were to use the default mean-field guide you'd get this for free, but you'd loose the ability to specify initial values for the params, so maybe go with a squishing function.

**Comment by JsonFreeman** [comment]
Oh I see! Yes, the variance going negative would certainly be a problem. But wouldn't that lead to a malformed distribution as opposed to the gradient being non-finite? I guess I am confused about the error message.

So the upshot is that when it is learning the variance, it overshoots when the step size is high and the value becomes negative. Is that accurate?

**Comment by null-a** [comment]
> But wouldn't that lead to a malformed distribution as opposed to the gradient being non-finite?

Yeah, when the variance goes negative, the score from the `Gaussian` in the guide will be `NaN`. `scoreDiff` is just the difference between the total score from the guide and the target for a particular execution.

(I agree the message could be better.)

> So the upshot is that when it is learning the variance, it overshoots when the step size is high and the value becomes negative. Is that accurate?

Yeah, except that I wouldn't attribute it to the step size been too large. Because we're using a monte-carlo estimate of the gradient, the noise in the estimate could cause an 'overshoot' even with a tiny step size.

**Comment by JsonFreeman** [comment]
I see. Thanks so much for looking into this Paul!

**Comment by null-a** [comment]
No problem!
